README.md
# Observatory extension

CartoDB extension that implements the row-level functions needed by the Observatory Service.

## Code organization

```
.
├── doc     # documentation
├── release # released versions
└── src     # source code
    └── pg
        ├── sql
        └── test
            ├── expected
            ├── fixtures
            └── sql
```

# Development workflow

We distinguish two roles regarding the development cycle:

* *developers* will implement new functionality and bugfixes into
  the codebase and will request for new releases of the extension.
* A *release manager* will attend these requests and will handle
  the release process. The release process is sequential:
  no concurrent releases will ever be in the works.

We use the default `develop` branch as the basis for development.
The `master` branch is used to merge and tag releases to be
deployed in production.

Developers shall create a new topic branch from `develop` for any new feature
or bugfix and commit their changes to it and eventually merge back into
the `develop` branch. When a new release is required a Pull Request
will be open against the `develop` branch.

The `develop` pull requests will be handled by the release manage,
who will merge into master where new releases are prepared and tagged.
The `master` branch is the sole responsibility of the release masters
and developers must not commit or merge into it.

## Development Guidelines

For a detailed description of the development process please see
the [CONTRIBUTING.md](CONTRIBUTING.md) guide.

Any modification to the source code
shall always be done in a topic branch created from the `develop` branch.

Tests, documentation and peer code reviews are required for all
modifications.

The tests are executed by running this from the top directory:
```
sudo make install
make test
```
## Release

The release and deployment process is described in the
[RELEASE.md](RELEASE.md) guide and it is the responsibility of the designated
release manager.


NEWS.md
1.10.0 (2018-07-??)
-------------------

__Improvements__

* Updated for PostgreSQL 12 and PostGIS 3.0 compatibility.

1.9.0 (2018-04-20)
------------------

__Improvements__

* Improved `OBS_GetAvailableGeometries` for the DO Timespans project ([#325](https://github.com/CartoDB/observatory-extension/pull/325))
* Improved `OBS_GetAvailableTimespans` for the DO Timespans project ([#324](https://github.com/CartoDB/bigmetadata/issues/324))
* Modified the denominated suggested_name to mitigate collisions ([#327](https://github.com/CartoDB/observatory-extension/pull/327))
* Fixed some errors so now the extension supports PostgreSQL 10 ([#329](https://github.com/CartoDB/observatory-extension/pull/329))
* Fixed documentation
* Add support for multiple PostgreSQL and Postgis versions in our travis script for test purposes

1.8.0 (2017-10-18)
------------------

__Improvements__

* Add `number_geometries` field to `OBS_GetAvailableGeometries` in order to provide the number of geometries from the source data to be used in the score calculation ([#313](https://github.com/CartoDB/observatory-extension/issues/313))

1.7.0 (2017-08-18)
------------------

__Improvements__

* Add Travis support to execute the extension tests ([#183](https://github.com/CartoDB/observatory-extension/issues/183))

__API Changes__

* Add new function `OBS_MetadataValidation` ([#303](https://github.com/CartoDB/observatory-extension/pull/303))

__Bugfixes__

* Fixed parentheses for obs_getdata with ids
* Fixed failing tests due changes in the data dump for some TIGER geometries

1.6.0 (2017-07-20)
------------------

__Improvements__

* The current OBS_GetAvailableNumerators is not designed with our
  UI in mind so it's causing a lot of troubles and we're doing so
  many hacks to fit our UI needs and the interface of the function so this
  function it's a better fit for our purposes. ([#300](https://github.com/CartoDB/observatory-extension/pull/300))
* Now use the new meta table `obs_meta_geom_numer_timespan` to filter
  the geometries by geometries timespan and/or numerator timespan (which
  is what we get when we use the obs_getavailabletimespans) ([#302](https://github.com/CartoDB/observatory-extension/pull/302))

__Bugfixes__

* Right now we're doing INNER JOINS when we JOIN the `_procgeoms` and
  the data so we end up with NULL value instead of id, NULL value. ([#298](https://github.com/CartoDB/observatory-extension/pull/298))


1.5.1 (2017-05-16)
------------------

__Improvements__

* Much improved performance for `OBS_GetData` when augmenting with several
  different geometries simultaneously ([#285](https://github.com/CartoDB/observatory-extension/pull/285))
* Return the automatically assigned normalization type from `OBS_GetMeta`
  ([#285](https://github.com/CartoDB/observatory-extension/pull/285))

1.5.0 (2017-04-24)
------------------

__API Changes__

* Add `suggested_name` to `OBS_GetMeta` responses
  ([#281](https://github.com/CartoDB/observatory-extension/pull/281))
* Add `geom_type`, `geom_extra`, and `geom_tags` to
  `OBS_GetAvailableGeometries`.  This brings it up to spec with existing docs.
  ([#282](https://github.com/CartoDB/observatory-extension/pull/282))
* Add `timespan_type`, `timespan_extra`, and `timespan_tags` to
  `OBS_GetAvailableTimespans` for consistency.
  ([#282](https://github.com/CartoDB/observatory-extension/pull/282))

1.4.0 (2017-03-21)
------------------

__API Changes__

* Allow for override of `target_area` and `target_geoms` in `OBS_GetMeta`
  ([#276](https://github.com/CartoDB/observatory-extension/pull/276)).  This
  allows the interface to work with points and sparse areas much btter.
* Allow for override of `max_timespan_rank` and `max_score_rank` on an
  item-by-item basis for metadata.
* `numer_description`, `geom_description`, `denom_description`,
  `numer_t_description`, `denom_t_description` and `geom_t_description` now
  returned as part of `OBS_GetMeta`.

__Improvements__

* Reduced amount of simplification done on input geometries (from 0.0001 above
  500 points to 0.00001 above 1000 points).
* Added tests to confirm that accurate results are returned from automatic
  boundary selection

1.3.5 (2017-03-15)
------------------

No changes.  Artifact to allow for data update.

1.3.4 (2017-03-10)
------------------

__Bugfixes__

* Remove erroneously committed `RAISE NOTICE` in `OBS_GetData`

1.3.3 (2017-03-10)
------------------

__Bugfixes__

* Resolve divide-by-zero errors in cases where the intersection of an
  Observatory geometry and user geometry has 0 area
  ([#265](https://github.com/CartoDB/observatory-extension/pull/265))
* Run MakeValid on geometry's when intersecting, if necessary
  ([#268](https://github.com/CartoDB/observatory-extension/pull/268))

__Improvements__

* Add performance tests for multiple columns in `OBS_GetData`
* Major performance boost for `autotest.py` through the use of multi-column
  `OBS_GetData` instead of separate `OBS_GetMeasure` calls for every single
  measurement.
  ([#268](https://github.com/CartoDB/observatory-extension/pull/268))
* Major performance boost for `OBS_GetData` in cases where multiple columns are
  requested.  Previously, each additional column would result in a linear
  slowdown, even if geometries could be reused.
  ([#267](https://github.com/CartoDB/observatory-extension/pull/267))

1.3.2 (2017-03-02)
------------------

__Bugfixes__

* Accept "prenormalized" as well as "predenominated" to bypass normalization.
  This fixes issues with Camshaft.

1.3.1 (2017-02-16)
------------------

__Improvements__

* It is now possible to obtain measures that are averages or medians over
  arbitrary polygons ([#254](https://github.com/CartoDB/observatory-extension/pull/254).
* Added test point for Australian data
* `OBS_GetLegacyMetadata` now returns median and averages in cases where it is
  called for measures for polygons

1.3.0 (2017-01-17)
------------------

__API Changes__


src/python/test/util.py
import os
import psycopg2

DB_CONN = psycopg2.connect('postgres://{user}:{password}@{host}:{port}/{database}'.format(
    user=os.environ.get('PGUSER', 'postgres'),
    password=os.environ.get('PGPASSWORD', ''),
    host=os.environ.get('PGHOST', 'localhost'),
    port=os.environ.get('PGPORT', '5432'),
    database=os.environ.get('PGDATABASE', 'postgres'),
))
CURSOR = DB_CONN.cursor()


def query(q):
    '''
    Query the database.
    '''
    try:
        CURSOR.execute(q)
        return CURSOR
    except:
        DB_CONN.rollback()
        raise


def commit():
    try:
        DB_CONN.commit()
    except:
        DB_CONN.rollback()
        raise


scripts/ci/docker-test.sh
#!/bin/bash

docker run -e PGHOST=localhost -e PGPORT=5432 -v `pwd`:/srv --entrypoint="/bin/bash" ${1} /srv/scripts/ci/run_tests_docker.sh && \
    docker ps --filter status=dead --filter status=exited -aq | xargs docker rm -v


src/python/test/autotest.py
from nose.tools import assert_equal, assert_is_not_none
from nose_parameterized import parameterized

from itertools import zip_longest
from util import query
from collections import OrderedDict
import json


def grouper(iterable, n, fillvalue=None):
    "Collect data into fixed-length chunks or blocks"
    # grouper('ABCDEFG', 3, 'x') --> ABC DEF Gxx
    args = [iter(iterable)] * n
    return zip_longest(fillvalue=fillvalue, *args)


USE_SCHEMA = True

SKIP_COLUMNS = set([
    'mx.inegi_columns.INDI18',
    'mx.inegi_columns.ECO40',
    'mx.inegi_columns.POB34',
    'mx.inegi_columns.POB63',
    'mx.inegi_columns.INDI7',
    'mx.inegi_columns.EDU28',
    'mx.inegi_columns.SCONY10',
    'mx.inegi_columns.EDU31',
    'mx.inegi_columns.POB7',
    'mx.inegi_columns.VIV30',
    'mx.inegi_columns.INDI12',
    'mx.inegi_columns.EDU13',
    'mx.inegi_columns.ECO43',
    'mx.inegi_columns.VIV9',
    'mx.inegi_columns.HOGAR25',
    'mx.inegi_columns.POB32',
    'mx.inegi_columns.ECO7',
    'mx.inegi_columns.INDI19',
    'mx.inegi_columns.INDI16',
    'mx.inegi_columns.POB65',
    'mx.inegi_columns.INDI3',
    'mx.inegi_columns.INDI9',
    'mx.inegi_columns.POB36',
    'mx.inegi_columns.POB33',
    'mx.inegi_columns.POB58',
    'mx.inegi_columns.DISC4',
    'mx.inegi_columns.VIV41',
    'mx.inegi_columns.VIV40',
    'mx.inegi_columns.VIV17',
    'mx.inegi_columns.VIV25',
    'mx.inegi_columns.EDU10',
    'whosonfirst.wof_disputed_name',
    'us.census.tiger.fullname',
    'whosonfirst.wof_marinearea_name',
    'us.census.tiger.mtfcc',
    'whosonfirst.wof_county_name',
    'whosonfirst.wof_region_name',
    'fr.insee.P12_RP_CHOS',
    'fr.insee.P12_RP_HABFOR',
    'fr.insee.P12_RP_EAUCH',
    'fr.insee.P12_RP_BDWC',
    'fr.insee.P12_RP_MIDUR',
    'fr.insee.P12_RP_CLIM',
    'fr.insee.P12_RP_MIBOIS',
    'fr.insee.P12_RP_CASE',
    'fr.insee.P12_RP_TTEGOU',
    'fr.insee.P12_RP_ELEC',
    'fr.insee.P12_ACTOCC15P_ILT45D',
    'fr.insee.P12_RP_CHOS',
    'fr.insee.P12_RP_HABFOR',
    'fr.insee.P12_RP_EAUCH',
    'fr.insee.P12_RP_BDWC',
    'fr.insee.P12_RP_MIDUR',
    'fr.insee.P12_RP_CLIM',
    'fr.insee.P12_RP_MIBOIS',
    'fr.insee.P12_RP_CASE',
    'fr.insee.P12_RP_TTEGOU',
    'fr.insee.P12_RP_ELEC',
    'fr.insee.P12_ACTOCC15P_ILT45D',
    'uk.ons.LC3202WA0007',
    'uk.ons.LC3202WA0010',
    'uk.ons.LC3202WA0004',
    'uk.ons.LC3204WA0004',
    'uk.ons.LC3204WA0007',
    'uk.ons.LC3204WA0010',
    'br.geo.subdistritos_name',
    # Problems with universe (denominator is zero in test area)
    'ca.statcan.census2016.cols_census.c0058_t',
    'ca.statcan.census2016.cols_census.c1878_f',
    'ca.statcan.census2016.cols_census.c1878_m',
    'ca.statcan.census2016.cols_census.c1674_t',
    'ca.statcan.census2016.cols_census.c1675_t',
    'ca.statcan.census2016.cols_census.c1676_t',
    'ca.statcan.census2016.cols_census.c1677_t',
    'ca.statcan.census2016.cols_census.c0801_t',
    'ca.statcan.census2016.cols_census.c0802_t',
    'ca.statcan.census2016.cols_census.c0803_t',
    'ca.statcan.census2016.cols_census.c0805_t',
    'ca.statcan.census2016.cols_census.c0806_t',
    'ca.statcan.census2016.cols_census.c0807_t',
    'ca.statcan.census2016.cols_census.c0809_t',
    'ca.statcan.census2016.cols_census.c0810_t',
    'ca.statcan.census2016.cols_census.c0811_t',
    'ca.statcan.census2016.cols_census.c0813_t',
    'ca.statcan.census2016.cols_census.c0814_t',
    'ca.statcan.census2016.cols_census.c0815_t',
    'ca.statcan.census2016.cols_census.c0817_t',
    'ca.statcan.census2016.cols_census.c0818_t',
    'ca.statcan.census2016.cols_census.c0820_t',
    'ca.statcan.census2016.cols_census.c0821_t',
    'ca.statcan.census2016.cols_census.c0821_t',
    'ca.statcan.census2016.cols_census.c0823_t',
    'ca.statcan.census2016.cols_census.c0824_t',
    'ca.statcan.census2016.cols_census.c0826_t',
    'ca.statcan.census2016.cols_census.c0827_t',
    'ca.statcan.census2016.cols_census.c0829_t',
    'ca.statcan.census2016.cols_census.c0830_t',
    'ca.statcan.census2016.cols_census.c0832_t',
    'ca.statcan.census2016.cols_census.c0833_t'
])

MEASURE_COLUMNS = query('''
SELECT cdb_observatory.FIRST(distinct numer_id) numer_ids,
       numer_aggregate,
       denom_reltype
FROM observatory.obs_meta
WHERE numer_weight > 0
  AND numer_id NOT IN ('{skip}')
  AND numer_id NOT LIKE 'eu.%' --Skipping Eurostat
  AND section_tags IS NOT NULL
  AND subsection_tags IS NOT NULL
GROUP BY numer_id, numer_aggregate, denom_reltype
'''.format(skip="', '".join(SKIP_COLUMNS))).fetchall()


def default_lonlat(column_id):
    '''
    Returns default test point for the column_id.
    '''
    if column_id == 'whosonfirst.wof_disputed_geom':
        return (76.57, 33.78)
    elif column_id == 'whosonfirst.wof_marinearea_geom':
        return (-68.47, 43.33)
    elif column_id.startswith('uk'):
        if 'WA' in column_id:
            return (51.46844551219723, -3.184833526611328)
        else:
            return (51.51461834694225, -0.08883476257324219)
    elif column_id.startswith('es'):
        return (42.8226119029222, -2.51141249535454)
    elif column_id.startswith('us.zillow'):
        return (28.3305906291771, -81.3544048197256)
    elif column_id.startswith('mx.'):
        return (19.41347699386547, -99.17019367218018)
    elif column_id.startswith('fr.'):
        return (48.860875144709475, 2.3613739013671875)
    elif column_id.startswith('ca.'):
        return (43.65594991256823, -79.37965393066406)
    elif (column_id.startswith('us.census.tiger.school_district_elementary') or
          column_id.startswith('us.census.tiger.school_district_secondary')):
        return (40.7025, -73.7067)
    elif column_id.startswith('us.census.'):
        return (28.3305906291771, -81.3544048197256)
    elif column_id.startswith('us.ihme.'):
        return (28.3305906291771, -81.3544048197256)
    elif column_id.startswith('us.bls.'):
        return (28.3305906291771, -81.3544048197256)
    elif column_id.startswith('us.qcew.'):
        return (28.3305906291771, -81.3544048197256)
    elif column_id.startswith('whosonfirst.'):
        return (28.3305906291771, -81.3544048197256)
    elif column_id.startswith('us.epa.'):
        return (28.3305906291771, -81.3544048197256)
    elif column_id.startswith('br.'):
        return (-23.53, -46.63)
    elif column_id.startswith('au.'):
        return (-33.8806, 151.2131)
    else:
        raise Exception('No catalog point set for {}'.format(column_id))


def default_point(test_point):
    lat, lng = test_point
    return 'ST_SetSRID(ST_MakePoint({lng}, {lat}), 4326)'.format(
        lat=lat, lng=lng)


def default_area(test_point):
    '''
    Returns default test area for the column_id
    '''
    point = default_point(test_point)
    area = 'ST_Transform(ST_Buffer(ST_Transform({point}, 3857), 250), 4326)'.format(
        point=point)
    return area


def filter_points():
    return MEASURE_COLUMNS


def filter_areas():
    filtered = []
    for numer_ids, numer_aggregate, denom_reltype in MEASURE_COLUMNS:
        if numer_aggregate is None or numer_aggregate.lower() not in ('sum', 'median', 'average'):
            continue
        if numer_aggregate.lower() in ('median', 'average') \
                and (denom_reltype is None or denom_reltype.lower() != 'universe'):
            continue
        filtered.append((numer_ids, numer_aggregate, denom_reltype))

    return filtered


def grouped_measure_columns(filtered_columns):
    groupbypoint = dict()
    for row in filtered_columns:
        numer_ids = row[0]
        point = default_lonlat(numer_ids)
        if point in groupbypoint:
            groupbypoint[point].append(numer_ids)
        else:
            groupbypoint[point] = [numer_ids]

    for point, numer_ids in groupbypoint.items():
        for colgroup in grouper(numer_ids, 50):
            yield point, [c for c in colgroup if c]


@parameterized(grouped_measure_columns(filter_points()))
def test_get_measure_points(point, numer_ids):
    _test_measures(numer_ids, default_point(point))


@parameterized(grouped_measure_columns(filter_areas()))
def test_get_measure_areas(point, numer_ids):
    _test_measures(numer_ids, default_area(point))


def _test_measures(numer_ids, geom):
    in_params = []
    for numer_id in numer_ids:
        in_params.append({
            'numer_id': numer_id,
            'normalization': 'predenominated'
        })

    params = query('''
        SELECT {schema}OBS_GetMeta({geom}, '{in_params}')
    '''.format(schema='cdb_observatory.' if USE_SCHEMA else '',
               geom=geom,
               in_params=json.dumps(in_params))).fetchone()[0]

    # We can get duplicate IDs from multi-denominators, so for now we
    # compress those measures into a single
    params = list(OrderedDict([(p['id'], p) for p in params]).values())
    assert_equal(len(params), len(in_params),
                 'Inconsistent out and in params for {}'.format(in_params))

    q = '''
    SELECT * FROM {schema}OBS_GetData(ARRAY[({geom}, 1)::geomval], '{params}')
    '''.format(schema='cdb_observatory.' if USE_SCHEMA else '',
               geom=geom,
               params=json.dumps(params).replace("'", "''"))
    resp = query(q).fetchone()
    assert_is_not_none(resp, 'NULL returned for {}'.format(in_params))
    rawvals = resp[1]
    vals = [v['value'] for v in rawvals]

    assert_equal(len(vals), len(in_params))
    for i, val in enumerate(vals):
        assert_is_not_none(val, 'NULL for {}'.format(in_params[i]['numer_id']))


scripts/ci/install_postgres.sh
#!/bin/bash

# echo commands
set -x

# exit on error
set -e

dpkg -l | grep postgresql

# Add the PDGD repository
apt-key adv --keyserver keys.gnupg.net --recv-keys 7FCC7D46ACCC4CF8
add-apt-repository "deb http://apt.postgresql.org/pub/repos/apt/ trusty-pgdg main"
apt-get update

# Remove those all PgSQL versions except the one we're testing
PGSQL_VERSIONS=(9.2 9.3 9.4 9.5 9.6 10)
/etc/init.d/postgresql stop # stop travis default instance
for V in "${PGSQL_VERSIONS[@]}"; do
    if [ "$V" != "$PGSQL_VERSION" ]; then
        apt-get -y remove --purge postgresql-${V} postgresql-client-${V} postgresql-contrib-${V} postgresql-${V}-postgis-2.3-scripts
    else
        apt-get -y remove --purge postgresql-${V}-postgis-2.3-scripts
    fi
done

apt-get -y autoremove

# Install PostgreSQL
apt-get -y install postgresql-${PGSQL_VERSION} postgresql-${PGSQL_VERSION}-postgis-${POSTGIS_VERSION} postgresql-server-dev-${PGSQL_VERSION} postgresql-plpython-${PGSQL_VERSION}

# Configure it to accept local connections from postgres
echo -e "# TYPE  DATABASE        USER            ADDRESS                 METHOD \nlocal   all             postgres                                trust\nlocal   all             all                                     trust\nhost    all             all             127.0.0.1/32            trust" > /etc/postgresql/${PGSQL_VERSION}/main/pg_hba.conf

# Restart PostgreSQL
/etc/init.d/postgresql restart ${PGSQL_VERSION}

dpkg -l | grep postgresql


scripts/autotest.py
from nose.tools import assert_equal, assert_is_not_none
from nose_parameterized import parameterized

import os
import re
import requests

HOSTNAME = os.environ['OBS_HOSTNAME']
API_KEY = os.environ['OBS_API_KEY']
META_HOSTNAME = os.environ.get('OBS_META_HOSTNAME', HOSTNAME)
META_API_KEY = os.environ.get('OBS_META_API_KEY', API_KEY)
USE_SCHEMA = 'OBS_USE_SCHEMA' in os.environ


def query(q, is_meta=False, **options):
    '''
    Query the account.  Returned is the response, wrapped by the requests
    library.
    '''
    url = 'https://{hostname}/api/v2/sql'.format(
        hostname=META_HOSTNAME if is_meta else HOSTNAME)
    params = options.copy()
    params['q'] = re.sub(r'\s+', ' ', q)
    params['api_key'] = META_API_KEY if is_meta else API_KEY
    return requests.get(url, params=params)

MEASURE_COLUMNS = [(r['numer_id'], r['point_only'], ) for r in query('''
SELECT distinct numer_id, numer_aggregate NOT ILIKE 'sum' as point_only
FROM obs_meta
WHERE numer_type ILIKE 'numeric'
AND numer_weight > 0
''', is_meta=True).json()['rows']]

CATEGORY_COLUMNS = [(r['numer_id'], ) for r in query('''
SELECT distinct numer_id
FROM obs_meta
WHERE numer_type ILIKE 'text'
AND numer_weight > 0
''', is_meta=True).json()['rows']]

BOUNDARY_COLUMNS = [(r['id'], ) for r in query('''
SELECT id FROM obs_column
WHERE type ILIKE 'geometry'
AND weight > 0
''', is_meta=True).json()['rows']]

US_CENSUS_MEASURE_COLUMNS = [(r['numer_name'], ) for r in query('''
SELECT distinct numer_name
FROM obs_meta
WHERE numer_type ILIKE 'numeric'
AND 'us.census.acs.acs' = ANY (subsection_tags)
AND numer_weight > 0
''', is_meta=True).json()['rows']]


def default_geometry_id(column_id):
    '''
    Returns default test point for the column_id.
    '''
    if column_id == 'whosonfirst.wof_disputed_geom':
        return 'CDB_LatLng(33.78, 76.57)'
    elif column_id == 'whosonfirst.wof_marinearea_geom':
        return 'CDB_LatLng(43.33, -68.47)'
    elif column_id in ('us.census.tiger.school_district_elementary',
                       'us.census.tiger.school_district_secondary',
                       'us.census.tiger.school_district_elementary_clipped',
                       'us.census.tiger.school_district_secondary_clipped'):
        return 'CDB_LatLng(40.7025, -73.7067)'
    elif column_id.startswith('es.ine'):
        return 'CDB_LatLng(42.8226119029222, -2.51141249535454)'
    elif column_id.startswith('us.zillow'):
        return 'CDB_LatLng(28.3305906291771, -81.3544048197256)'
    else:
        return 'CDB_LatLng(40.7, -73.9)'


def default_point(column_id):
    '''
    Returns default test point for the column_id.
    '''
    if column_id == 'whosonfirst.wof_disputed_geom':
        return 'CDB_LatLng(33.78, 76.57)'
    elif column_id == 'whosonfirst.wof_marinearea_geom':
        return 'CDB_LatLng(43.33, -68.47)'
    elif column_id in ('us.census.tiger.school_district_elementary',
                       'us.census.tiger.school_district_secondary',
                       'us.census.tiger.school_district_elementary_clipped',
                       'us.census.tiger.school_district_secondary_clipped'):
        return 'CDB_LatLng(40.7025, -73.7067)'
    elif column_id.startswith('uk'):
        if 'WA' in column_id:
            return 'CDB_LatLng(51.46844551219723, -3.184833526611328)'
        else:
            return 'CDB_LatLng(51.51461834694225, -0.08883476257324219)'
    elif column_id.startswith('es'):
        return 'CDB_LatLng(42.8226119029222, -2.51141249535454)'
    elif column_id.startswith('us.zillow'):
        return 'CDB_LatLng(28.3305906291771, -81.3544048197256)'
    elif column_id.startswith('mx.'):
        return 'CDB_LatLng(19.41347699386547, -99.17019367218018)'
    else:
        return 'CDB_LatLng(40.7, -73.9)'


def default_area(column_id):
    '''
    Returns default test area for the column_id
    '''
    point = default_point(column_id)
    area = 'ST_Transform(ST_Buffer(ST_Transform({point}, 3857), 1000), 4326)'.format(
        point=point)
    return area

@parameterized(US_CENSUS_MEASURE_COLUMNS)
def test_get_us_census_measure_points(name):
    print 'test_get_us_census_measure_points, ', name
    resp = query('''
SELECT * FROM {schema}OBS_GetUSCensusMeasure({point}, '{name}')
                 '''.format(name=name.replace("'", "''"),
                            schema='cdb_observatory.' if USE_SCHEMA else '',
                            point=default_point('')))
    assert_equal(resp.status_code, 200)
    rows = resp.json()['rows']
    assert_equal(1, len(rows))
    assert_is_not_none(rows[0].values()[0])


@parameterized(MEASURE_COLUMNS)
def test_get_measure_areas(column_id, point_only):
    print 'test_get_measure_areas, ', column_id, point_only
    if point_only:
        return
    resp = query('''
SELECT * FROM {schema}OBS_GetMeasure({area}, '{column_id}')
                 '''.format(column_id=column_id,
                            schema='cdb_observatory.' if USE_SCHEMA else '',
                            area=default_area(column_id)))
    assert_equal(resp.status_code, 200)
    rows = resp.json()['rows']
    assert_equal(1, len(rows))
    assert_is_not_none(rows[0].values()[0])


@parameterized(MEASURE_COLUMNS)
def test_get_measure_points(column_id, point_only):
    print 'test_get_measure_points, ', column_id, point_only
    resp = query('''
SELECT * FROM {schema}OBS_GetMeasure({point}, '{column_id}')
                 '''.format(column_id=column_id,
                            schema='cdb_observatory.' if USE_SCHEMA else '',
                            point=default_point(column_id)))
    assert_equal(resp.status_code, 200)
    rows = resp.json()['rows']
    assert_equal(1, len(rows))
    assert_is_not_none(rows[0].values()[0])

#@parameterized(CATEGORY_COLUMNS)
#def test_get_category_areas(column_id):
#    resp = query('''
#SELECT * FROM {schema}OBS_GetCategory({area}, '{column_id}')
#                 '''.format(column_id=column_id,
#                            schema='cdb_observatory.' if USE_SCHEMA else '',
#                            area=default_area(column_id)))
#    assert_equal(resp.status_code, 200)
#    rows = resp.json()['rows']
#    assert_equal(1, len(rows))
#    assert_is_not_none(rows[0].values()[0])

@parameterized(CATEGORY_COLUMNS)
def test_get_category_points(column_id):
    print 'test_get_category_points, ', column_id
    resp = query('''
SELECT * FROM {schema}OBS_GetCategory({point}, '{column_id}')
                 '''.format(column_id=column_id,
                            schema='cdb_observatory.' if USE_SCHEMA else '',
                            point=default_point(column_id)))
    assert_equal(resp.status_code, 200)
    rows = resp.json()['rows']
    assert_equal(1, len(rows))
    assert_is_not_none(rows[0].values()[0])

#@parameterized(BOUNDARY_COLUMNS)
#def test_get_boundaries_by_geometry(column_id):
#    resp = query('''
#SELECT * FROM {schema}OBS_GetBoundariesByGeometry({area}, '{column_id}')
#                 '''.format(column_id=column_id,
#                            schema='cdb_observatory.' if USE_SCHEMA else '',
#                            area=default_area(column_id)))
#    assert_equal(resp.status_code, 200)
#    rows = resp.json()['rows']
#    assert_equal(1, len(rows))
#    assert_is_not_none(rows[0].values()[0])

#@parameterized(BOUNDARY_COLUMNS)
#def test_get_points_by_geometry(column_id):
#    resp = query('''
#SELECT * FROM {schema}OBS_GetPointsByGeometry({area}, '{column_id}')
#                 '''.format(column_id=column_id,
#                            schema='cdb_observatory.' if USE_SCHEMA else '',
#                            area=default_area(column_id)))
#    assert_equal(resp.status_code, 200)
#    rows = resp.json()['rows']
#    assert_equal(1, len(rows))
#    assert_is_not_none(rows[0].values()[0])

#@parameterized(BOUNDARY_COLUMNS)
#def test_get_boundary_points(column_id):
#    resp = query('''
#SELECT * FROM {schema}OBS_GetBoundary({point}, '{column_id}')
#                 '''.format(column_id=column_id,
#                            schema='cdb_observatory.' if USE_SCHEMA else '',
#                            point=default_point(column_id)))
#    assert_equal(resp.status_code, 200)
#    rows = resp.json()['rows']
#    assert_equal(1, len(rows))
#    assert_is_not_none(rows[0].values()[0])

#@parameterized(BOUNDARY_COLUMNS)
#def test_get_boundary_id(column_id):
#    resp = query('''
#SELECT * FROM {schema}OBS_GetBoundaryId({point}, '{column_id}')
#                 '''.format(column_id=column_id,
#                            schema='cdb_observatory.' if USE_SCHEMA else '',
#                            point=default_point(column_id)))
#    assert_equal(resp.status_code, 200)
#    rows = resp.json()['rows']
#    assert_equal(1, len(rows))
#    assert_is_not_none(rows[0].values()[0])

#@parameterized(BOUNDARY_COLUMNS)
#def test_get_boundary_by_id(column_id):
#    resp = query('''
#SELECT * FROM {schema}OBS_GetBoundaryById({geometry_id}, '{column_id}')
#                 '''.format(column_id=column_id,
#                            schema='cdb_observatory.' if USE_SCHEMA else '',
#                            geometry_id=default_geometry_id(column_id)))
#    assert_equal(resp.status_code, 200)
#    rows = resp.json()['rows']
#    assert_equal(1, len(rows))
#    assert_is_not_none(rows[0].values()[0])
